hoodie-common/src/main/java/com/uber/hoodie/common/table/log/block/HoodieDeleteBlock.java

```diff
@@ -16,64 +16,100 @@
 
+import com.uber.hoodie.common.model.HoodieLogFile;
+import com.uber.hoodie.common.storage.SizeAwareDataInputStream;
+import com.uber.hoodie.exception.HoodieIOException;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.fs.FSDataInputStream;
+
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.DataInputStream;
 import java.io.DataOutputStream;
 import java.io.IOException;
 import java.nio.charset.Charset;
+import java.util.HashMap;
 import java.util.Map;
-
-import com.uber.hoodie.common.storage.SizeAwareDataInputStream;
-import org.apache.commons.lang3.StringUtils;
+import java.util.Optional;
 
 /**
  * Delete block contains a list of keys to be deleted from scanning the blocks so far
  */
 public class HoodieDeleteBlock extends HoodieLogBlock {
 
-  private final String[] keysToDelete;
+  private String[] keysToDelete;
 
-  public HoodieDeleteBlock(String[] keysToDelete, Map<LogMetadataType, String> metadata) {
-    super(metadata);
+  public HoodieDeleteBlock(String[] keysToDelete,
+                           Map<HeaderMetadataType, String> header) {
+    this(Optional.empty(), null, false, Optional.empty(), header, new HashMap<>());
     this.keysToDelete = keysToDelete;
   }
 
-  public HoodieDeleteBlock(String[] keysToDelete) {
-    this(keysToDelete, null);
+
+  private HoodieDeleteBlock(Optional<byte[]> content, FSDataInputStream inputStream,
+                            boolean readBlockLazily, Optional<HoodieLogBlockContentLocation> blockContentLocation,
+                            Map<HeaderMetadataType, String> header, Map<HeaderMetadataType, String> footer) {
+    super(header, footer, blockContentLocation, content, inputStream, readBlockLazily);
   }
 
   @Override
-  public byte[] getBytes() throws IOException {
+  public byte[] getContentBytes() throws IOException {
+
+    // In case this method is called before realizing keys from content
+    if (getContent().isPresent()) {
+      return getContent().get();
+    } else if (readBlockLazily && !getContent().isPresent() && keysToDelete == null) {
+      // read block lazily
+      getKeysToDelete();
+    }
+
     ByteArrayOutputStream baos = new ByteArrayOutputStream();
     DataOutputStream output = new DataOutputStream(baos);
-    if (super.getLogMetadata() != null) {
-      output.write(HoodieLogBlock.getLogMetadataBytes(super.getLogMetadata()));
-    }
-    byte[] bytesToWrite = StringUtils.join(keysToDelete, ',').getBytes(Charset.forName("utf-8"));
+    byte[] bytesToWrite = StringUtils.join(getKeysToDelete(), ',').getBytes(Charset.forName("utf-8"));
+    output.writeInt(HoodieLogBlock.version);
     output.writeInt(bytesToWrite.length);
     output.write(bytesToWrite);
     return baos.toByteArray();
   }
 
   public String[] getKeysToDelete() {
-    return keysToDelete;
+    try {
+      if (keysToDelete == null) {
+        if (!getContent().isPresent() && readBlockLazily) {
```

Comment:
same here.. its a general comment on these blocks.. there is a reader level switch already that chooses to read or skip content, so why should the blocks themselves be aware?

Reply:
Following is the reason to keep this check (notice a similar check in HoodieAvroDataBlock) as well : 

1) If lazy reading is turned off, all the content is read in memory and then immediately either converted to List<IndexedRecords> or List<Keys> in this case. In HoodieScanner, we will keep multiple of such blocks in memory (for rollback etc) and this causes OOM. So this change is to push down the conversion of byte [] to the specific data.

Reply:
still have my doubts here.. lets reserve it for follow up again..

Reply:
I'll note this and sync up with you in person.